src/NServiceBus.Core.Tests/Pipeline/TestableMessageOperations.cs

```diff
@@ -0,0 +1,35 @@
+﻿namespace NServiceBus.Core.Tests.Pipeline
+{
+    using System;
+    using NServiceBus.Pipeline;
+    using System.Threading.Tasks;
+    using MessageInterfaces.MessageMapper.Reflection;
+
+    class TestableMessageOperations : MessageOperations
+    {
+        public Pipeline<IOutgoingPublishContext> PublishPipeline => (Pipeline<IOutgoingPublishContext>)publishPipeline;
+        public Pipeline<IOutgoingSendContext> SendPipeline => (Pipeline<IOutgoingSendContext>)sendPipeline;
+        public Pipeline<IOutgoingReplyContext> ReplyPipeline => (Pipeline<IOutgoingReplyContext>)replyPipeline;
+        public Pipeline<ISubscribeContext> SubscribePipeline => (Pipeline<ISubscribeContext>)subscribePipeline;
+        public Pipeline<IUnsubscribeContext> UnsubscribePipeline => (Pipeline<IUnsubscribeContext>)unsubscribePipeline;
+
+        public TestableMessageOperations() : base(new MessageMapper(), new Pipeline<IOutgoingPublishContext>(), new Pipeline<IOutgoingSendContext>(), new Pipeline<IOutgoingReplyContext>(), new Pipeline<ISubscribeContext>(), new Pipeline<IUnsubscribeContext>(), new ActivityFactory())
+        {
+        }
+
+        public class Pipeline<TContext> : IPipeline<TContext> where TContext : IBehaviorContext
+        {
+            public Action<TContext> OnInvoke { get; set; }
```

Comment:
A wild thought: are we mostly interested in the context when things are invoked? Then maybe we can simplify this by directly expose a settable action delegate instead of going over the pipeline into the action delegate

But that might complicate the tests when you only want to access the context like you do in some. I might have an idea but I'm not sure if it is simpler

Reply:
> A wild thought: are we mostly interested in the context when things are invoked? Then maybe we can simplify this by directly expose a settable action delegate instead of going over the pipeline into the action delegate

not sure I can follow. Can you explain your thoughts further?

Reply:
Well first I was thinking about just having dedicated properties directly on the testable implementation instead of nesting them further. So instead of doing

```csharp
testableOperations.SendPipeline.Context
```

you would do

```csharp
testableOperations.SentContext
```

or at it the simplest form only support the delegates like

```csharp
testableOperations.OnSend = (ctx) => { }
testableOperations.OnPublish = (ctx) => { }
```

Because I was also thinking having a settable property called `Context` that is overridden by each invocation could be misleadingly simply and lead to problems down the line. I'm aware that previous implementations had similar problems but they were only localized and now since we moved things into a common abstraction I guess we should think about those scenarios. By having the delegates only it forces you to override the delegate always and capture things based on the knowledge of the test (basically coding the assumptions of one or multiple invocations explicitly). Otherwise, from a generic "testable" implementation perspective we would need to expose a collection of contexts :(

Reply:
`testableOperations.OnSend = (ctx) => { }` I kinda like this idea, although I was also thinking about just renaming the `Context` property to something like `LastContext` to make the issue you described more clear anyway.

Reply:
I looked a bit more into this, while I really like the `testableOperations.OnSend = (ctx) => { }` suggestion, we seem to use the context in quite a few places so this change would require significantly more effort to capture the context, so I think the current approach is a bit more useful. I renamed context to `LastContext` to indicate a bit more clearly that it will be overridden on multiple calls. I'm aware this is not the most elegant solution but as a testing helper, I think this works well enough for now.

Reply:
Cool. Makes sense. Thanks for considering